### PR TITLE
be strict about binding names

### PR DESCRIPTION
**Pull request by jameinel**
## Description of change

When deploying an application, you're allowed to bind endpoints to spaces.
However, we would allow any endpoint name to be supplied, which meant that
typos would not be caught.
(https://pad.lv/1661929)

Further, when deploying an application and only supplying the default binding,
if that application did not have any relations at all, it wouldn't be put
into the preferred space. We now record the default binding request along
with the other endpoint bindings. (https://pad.lv/1654952)

## QA steps

For the first half, use a provider that supports spaces (MAAS or AWS), and try to deploy an application but supply an invalid endpoint name. For example:
```
  $ juju deploy mysql --bind NOTdb=space
```
The old Juju would silently ignore your request, the new Juju will give an error that "NOTdb" is not a valid binding.

For the second half, find a charm that doesn't have any relations (ubuntu being the canonical example), and then try to deploy it into a space. This is most obvious if you deploy into a container on a host that has multiple spaces
```
  $ juju deploy ubuntu --bind space --to lxd:0
```
The old Juju would treat it as though you supplied nothing, rather than trying to allocate a machine with access to 'space'.

## Documentation changes

This only gives errors in cases where you were using the CLI incorrectly. I'm not sure we need explicit documentation of that. It is possible that some bundles that used to deploy will start failing, because they were not doing what they thought they were doing anyway.

## Bug reference

[lp:1661929](https://pad.lv/1661929)
[lp:1654952](https://pad.lv/1654952)

### CODE REVIEW

File: juju/deploy.go [hunk 1]
```diff
@@ -87,17 +93,58 @@ func DeployApplication(st ApplicationDeployer, args DeployApplicationParams) (*s
...
 	// empty space names as values, so we use defaultBindings as fallback.
 	defaultBindings := state.DefaultEndpointBindingsForCharm(charmMeta)
 	if givenBindings == nil {
 		givenBindings = make(map[string]string, len(defaultBindings))
 	}
+	if err := validateGivenBindings(givenBindings, defaultBindings); err != nil {
+		return nil, err
+	}
 
 	// Get the application-level default binding for all unspecified endpoint, if
```
Comment (wallyworld) [inline juju/deploy.go]: typo - endpoints

File: juju/deploy_test.go [hunk 2]
```diff
@@ -195,6 +198,61 @@ func (s *DeployLocalSuite) TestDeployWithBoundRelationNamesAndExtraBindingsNames
...
+			ApplicationName: "bob",
+			Charm:           wordpressCharm,
+			EndpointBindings: map[string]string{
+				"":   "public",
+				"db": "intranel", //typo 'intranel'
+			},
+		})
+	c.Assert(err, gc.ErrorMatches, `cannot add application "bob": unknown space "intranel" not valid`)
+	c.Check(service, gc.IsNil)
+	// The service should not have been added
```
Comment (wallyworld) [inline juju/deploy_test.go]: s/service/application

File: juju/deploy_test.go [hunk 3]
```diff
@@ -195,6 +198,61 @@ func (s *DeployLocalSuite) TestDeployWithBoundRelationNamesAndExtraBindingsNames
...
+			EndpointBindings: map[string]string{
+				"":      "public",
+				"bd":    "internal", // typo 'bd'
+				"pubic": "public",   // typo 'pubic'
+			},
+		})
+	c.Assert(err, gc.ErrorMatches,
+		`invalid binding\(s\) supplied "bd", "pubic", valid binding names are "admin-api", .* "url"`)
+	c.Check(service, gc.IsNil)
+	// The service should not have been added
```
Comment (wallyworld) [inline juju/deploy_test.go]: ditto

File: juju/deploy.go (hunk 1 above)
Comment (jameinel) [inline juju/deploy.go]: :+1:

File: juju/deploy_test.go (hunk 2 above)
Comment (jameinel) [inline juju/deploy_test.go]: 👍

File: juju/deploy_test.go (hunk 3 above)
Comment (jameinel) [inline juju/deploy_test.go]: 👍

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**